### ISSUE TITLE
Sidekiq 4.0 needs a strong warning regarding the use of Celluloid

### ISSUE THREAD

**Issue by leopoldodonnell** [mperham/sidekiq#2722]
We've been enjoying Sidekiq for several years now and were excited to upgrade to 4.0 with 3.0 Pro. As you can imagine a major release, as awesome as it is, can come with unexpected surprises...

**Our Story**

Wen upgrading we noted that Celluloid had been dropped, but mperham was still a Celluloid lover so we didn't think anything of it when we upgraded. All of our issues had been settled and everything appeared to pass, so we pushed to production.

Sadly, by the time we woke up the next day the logs were spewing tons of the following:

    ERROR: SessionEventProcessor:perform(4456838, webcam_failure) failed with error: could not obtain a database connection within 5 seconds (waited 5.000141964 seconds). The max pool size is currently 81; consider increasing it.

It was immediately clear that we had a connection leak, so we watched the logs and bounced the servers while we tracked down the issue and solution.

That's when we recalled the Celluloid issue...

We had a Worker base class that we used for Cron/Timer based workers and any of these types of Worker had Celluloid mixed in. So, every time one of our cron activities would fire, we would leak a connection. At least that was our suspicion looking at our database connection count etc.

As awful as a production fire-drill is, the solution was simple enough - we simply made a helper class that had the Celluloid code we required so that we no longer had to worry about the conflict.

----

**Bottom Line** Please put a strong notice that Workers should either avoid deriving from or mixing in Celluloid unless you plan to manage your own database connections.

**Comment by seuros** [comment]
Thank you, could you update the wiki.

**Comment by mperham** [comment]
It's the first line in the 4.0 upgrade notes.  How could I make it any plainer?

https://github.com/mperham/sidekiq/blob/master/4.0-Upgrade.md#whats-new
